src/string_map.c

```diff
@@ -147,46 +172,34 @@ rcutils_string_map_reserve(rcutils_string_map_t * string_map, size_t capacity)
     return RCUTILS_RET_OK;
   } else if (capacity == 0) {
     // if the requested capacity is zero, then make sure the existing keys and values are free'd
-    allocator.deallocate(string_map->impl->keys, allocator.state);
-    string_map->impl->keys = NULL;
-    allocator.deallocate(string_map->impl->values, allocator.state);
-    string_map->impl->values = NULL;
+    allocator.deallocate(string_map->impl->key_value_pairs, allocator.state);
```

Comment:
Note that I'm pretty sure we need the call to `__clear_string_map` here.  Otherwise, while we will free the list of key/value pairs, we won't actually free the keys and values themselves.

Reply:
It cannot be that `size != 0` and `capacity == 0`, because of the recursive call above. See https://github.com/ApexAI/rcutils/pull/1#discussion_r1120097853. So there are no keys and values to free.

Reply:
Oh, good point.  OK, this looks good to me then!